tensorboard/webapp/metrics/views/card_renderer/scalar_card_container.ts

```diff
@@ -332,16 +342,84 @@ export class ScalarCardContainer implements CardRenderer, OnInit {
               x: point.x - firstPoint.x,
             }));
           }
-          return {id: runId, points};
+          return {id: runId, runId, points};
         });
+      }),
+      combineLatestWith(
+        this.store.select(getMetricsScalarPartitionNonMonotonicX)
+      ),
+      map<
+        [
+          Array<{id: string; runId: string; points: ScalarCardPoint[]}>,
+          boolean
+        ],
+        LineChartDatum[]
+      >(([runsData, partitionX]) => {
+        if (!partitionX) {
+          return runsData.map((runData) => {
+            return {
+              ...runData,
+              partitionIndex: 0,
+              partitionSize: 1,
+            };
+          });
+        }
 
-        if (smoothing === 0) {
-          return of(dataSeriesList);
+        const partitionedRunsData: LineChartDatum[] = [];
+        for (const datum of runsData) {
+          const currentPartition: Array<Omit<
+            LineChartDatum,
+            'partitionSize' | 'partitionIndex'
+          >> = [];
+          let lastXValue = datum.points[0]?.x ?? 0;
+          let currentPoints: ScalarCardPoint[] = [];
+
+          for (const point of datum.points) {
+            if (point.x < lastXValue) {
+              currentPartition.push({
+                id: JSON.stringify([datum.id, currentPartition.length]),
+                runId: datum.runId,
+                points: currentPoints,
+              });
+              currentPoints = [];
+            }
+            currentPoints.push(point);
+            lastXValue = point.x;
+          }
+
+          currentPartition.push({
+            id:
+              currentPartition.length === 0
+                ? datum.id
+                : JSON.stringify([datum.id, currentPartition.length]),
+            runId: datum.runId,
+            points: currentPoints,
+          });
+
+          for (let index = 0; index < currentPartition.length; index++) {
+            partitionedRunsData.push({
+              ...currentPartition[index],
+              partitionIndex: index,
+              partitionSize: currentPartition.length,
+            });
+          }
+        }
+        return partitionedRunsData;
+      }),
+      shareReplay(1)
+    );
+
+    this.dataSeries$ = partitionedRuns$.pipe(
+      combineLatestWith(this.store.select(getMetricsScalarSmoothing)),
+      switchMap(([runsData, smoothing]) => {
+        const cleanedRunsData = runsData.map(({id, points}) => ({id, points}));
```

Comment:
The variable naming is starting to get really complex, imo.  If `partitionedRunsData` is a list, a name like shouldn't it contain `partitionedRunsDatum`s?

L327 `runsData` is a `{runId: string; points: ScalarCardPoint[]}`
L357 `runsData` is a `{id: string; runId: string; points: ScalarCardPoint[]}`
L370 `currentPartition` is a `{id: string; runId: string; points: ScalarCardPoint[]}`
L415 `cleanedRunsData` is a `{id: string, points: ScalarCardPoint[]}[]`
L441 `runIdAndPoints` is a `LineChartDatum`

iiuc, there are 5 different forms of "cleaning" that should occur, and more than one could be interpreted as "cleanedRunsData":
{runId,points} =>
{runId,pointsWithWallTime} =>
{runId,pointsWithWallTime,partitionIndex/size} ==>
{id,runId,(un)smoothedPointsWithWallTime,partitionIndex/size} ==>
{id,runId,(un)smoothedPointsWithWallTime}

It is fine with me to use the same name to refer to more than one of those 5 types, but could we make the names overall more consistent?